### PR TITLE
[AutoScheduler] Python based measure callbacks

### PR DESCRIPTION
**Pull request by comaniac**
The current auto_scheduler callbacks can only be implemented in C++. This PR exposes the interface of measure callbacks to Python so that people can plug in their own callback functions easily.

~~Note: looks like I cannot pass the `SearchPolicy` node to the `PackedFunc`. Specifically, `callback_func(policy, inputs, results);` results in the type mismatching error. This also prevents us from introducing the Python API for `SearchCallback`. Any advise is appreciated.~~

cc @merrymercy @jcf94

### CODE REVIEW

**Comment by comaniac** [comment]
Per offline discussion, we cast the abstract SearchPolicy to the actual instance so that we can pass it to the packed function.

@merrymercy @jcf94 PTAL.

**Review by jcf94 — APPROVED** [review]
LGTM. Thanks!

File: tests/python/relay/test_auto_scheduler_tuning.py [hunk 1]
```diff
@@ -23,6 +23,15 @@
 from test_auto_scheduler_task_extraction import get_network
 
 
+class CustomMeasureCallback(auto_scheduler.measure.PythonBasedMeasureCallback):
```
Comment (merrymercy) [inline tests/python/relay/test_auto_scheduler_tuning.py:26]: It is better to move this test to here (https://github.com/apache/tvm/blob/e51bcdd1ed99fc813454c68911f8032c852f48b4/tests/python/unittest/test_auto_scheduler_search_policy.py#L71).

Because this measure callback is not related to relay. We should not put it under `tests/python/relay/`

Comment (comaniac) [reply tests/python/relay/test_auto_scheduler_tuning.py:26]: Moved.

**Comment by comaniac** [comment]
I also took a look at the `SearchCallback` (it should be named to something like `PreSearchCallback` because it will be called once in the beginning of search instead of every round) but decided not to add a Python interface in this PR. Since we don't expose useful SearchPolicy APIs to Python, the Python-based callback can do nothing at this moment. Meanwhile, it's worthwhile to consider `PreloadCustomSketchRule` as @antinucleon suggested in another PR.

**Comment by jcf94** [comment]
> I also took a look at the `SearchCallback` (it should be named to something like `PreSearchCallback` because it will be called once in the beginning of search instead of every round) but decided not to add a Python interface in this PR. Since we don't expose useful SearchPolicy APIs to Python, the Python-based callback can do nothing at this moment. Meanwhile, it's worthwhile to consider `PreloadCustomSketchRule` as @antinucleon suggested in another PR.

Actually we already have a `init_search_callbacks` in the constructor function of SearchPolicy which will be called before the search process. And we also have the `PreloadCustomSketchRule` in our dev repo ...

**Comment by comaniac** [comment]
> Actually we already have a `init_search_callbacks` in the constructor function of SearchPolicy which will be called before the search process. And we also have the `PreloadCustomSketchRule` in our dev repo ...

Yeah that's why I bring up this, but my point is these callbacks are C-based instead of Python so there are not that convenient for users to experiment.

**Comment by jcf94** [comment]
> > Actually we already have a `init_search_callbacks` in the constructor function of SearchPolicy which will be called before the search process. And we also have the `PreloadCustomSketchRule` in our dev repo ...
> 
> Yeah that's why I bring up this, but my point is these callbacks are C-based instead of Python so there are not that convenient for users to experiment.

Oh, I find that comment: https://github.com/apache/tvm/pull/7132#discussion_r547936669
I agree with your point.
So seems we will need to:
1. Add the `PreloadCustomSketchRule` which enables users to add their own custom sketches. (And that's what we have already planed before but didn't take action to bring that here from our dev repo ....)
2. Just like your have added a python based `MeasureCallback` here, we'd better add a python based `SearchCallback` for init_search_callbacks.
3. Better re-arrange the `rules->push_back(.......)` part, maybe also move them to Python as a `InitSearchCallbak`.

Also cc @merrymercy @antinucleon

File: tests/python/relay/test_auto_scheduler_tuning.py [hunk 2]
```diff
@@ -23,6 +23,15 @@
...
+    """A simple Python-based callback for testing."""
+
+    def callback(self, policy, inputs, results):
+        assert isinstance(policy, auto_scheduler.search_policy.SketchPolicy)
+        for inp, res in zip(inputs, results):
+            print(policy, inp, res)
+
+
```
Comment (merrymercy) [inline tests/python/relay/test_auto_scheduler_tuning.py:34]:
```suggestion
```
Delete useless old code

**Review by merrymercy — APPROVED** [review]
